File: iOS7bookExamples/bk1ch12p350NotificationLeaker/ch12p325NotificationLeaker/FlipsideViewController.m

/*
Run the app and flip and flip back.
If dealloc is never called, you're leaking.
*/

#import "FlipsideViewController.h"

#define WHICH 0 // 0 shows the problem; solutions are 1, 2, 3 (and 4, but I don't advise it)


#if WHICH==0 //==================

// THE PROBLEM

@interface FlipsideViewController ()
@property (nonatomic, strong) id observer;
@end

@implementation FlipsideViewController

- (void) viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    NSLog(@"register %d", WHICH);
    self.observer = [[NSNotificationCenter defaultCenter]
                     addObserverForName:@"woohoo"
                     object:nil queue:nil
                     usingBlock:^(NSNotification *note)
                     {
                         [self description]; // potential leak
                     }];
}

- (void) viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    NSLog(@"%@", @"unregister");
    [[NSNotificationCenter defaultCenter] removeObserver:self.observer];
}

- (void) dealloc {
    // dealloc is never called, we are leaking
    NSLog(@"%@", @"dealloc");
}

#elif WHICH==1 //=================

// SOLUTION 1: weak ivar
// still can't remove observer in dealloc...
// because we have a retain cycle until we remove the observer!
// but at least removing the observer breaks the retain cycle and prevents the leak

@interface FlipsideViewController ()
@property (nonatomic, weak) id observer; // <=====
@end

@implementation FlipsideViewController

- (void) viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    NSLog(@"register %d", WHICH);
    self.observer = [[NSNotificationCenter defaultCenter]
                     addObserverForName:@"woohoo"
                     object:nil queue:nil
                     usingBlock:^(NSNotification *note)
                     {
                         [self description]; // potential leak
                     }];
}

- (void) viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    NSLog(@"%@", @"unregister");
    [[NSNotificationCenter defaultCenter] removeObserver:self.observer];
}

- (void) dealloc {
    NSLog(@"%@", @"dealloc");
}

#elif WHICH==2 //=================

// SOLUTION 2: explicit release of observer
// still can't remove observer in dealloc...
// because we have a retain cycle until we remove the observer!
// but at least removing the observer breaks the retain cycle and prevents the leak

@interface FlipsideViewController ()
@property (nonatomic, strong) id observer;
@end

@implementation FlipsideViewController

- (void) viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    NSLog(@"register %d", WHICH);
    self.observer = [[NSNotificationCenter defaultCenter]
                     addObserverForName:@"woohoo"
                     object:nil queue:nil
                     usingBlock:^(NSNotification *note)
                     {
                         [self description]; // potential leak
                     }];
}

- (void) viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    NSLog(@"%@", @"unregister");
    [[NSNotificationCenter defaultCenter] removeObserver:self.observer];
    self.observer = nil; // <=======
}

- (void) dealloc {
    NSLog(@"%@", @"dealloc");
}

#elif WHICH==3 //=================

// SOLUTION 3: weak-strong dance

@interface FlipsideViewController ()
@property (nonatomic, strong) id observer;
@end

@implementation FlipsideViewController  {
    NSString* myIvar;
}

- (void) viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    NSLog(@"register %d", WHICH);
    __weak id weakself = self; // <====
    self.observer = [[NSNotificationCenter defaultCenter]
                     addObserverForName:@"woohoo"
                     object:nil queue:nil
                     usingBlock:^(NSNotification *note)
                     {
                         FlipsideViewController* sself = weakself; // <===
                         if (sself) {                              // <===
                             [sself description]; // no leak - be careful not to mention self!
                             // do not use NSAssert!
                             // NSAssert(1, @"dummy assertion");
                             // do not use ivar names!
                             // myIvar = @"Testing"; // fortunately, compiler warns on this one
                         }
                     }];
}

- (void) dealloc {
    NSLog(@"%@", @"dealloc");
    // this solution has a big advantage: we can unregister in dealloc!
    NSLog(@"%@", @"unregister");
    [[NSNotificationCenter defaultCenter] removeObserver:self->_observer];
}

#elif WHICH==4 //=================

// SOLUTION 4: unregister in the block
// obviously this only works if you expect the block to be called exactly once
// but I don't like this approach - if the notification is never posted, we never unregister...
// and we are never released, so we leak; it's an unnecessary gamble

@interface FlipsideViewController ()
@end

@implementation FlipsideViewController

- (void) viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    NSLog(@"register %d", WHICH);
    __weak __block id observer = [[NSNotificationCenter defaultCenter]
                     addObserverForName:@"woohoo"
                     object:nil queue:nil
                     usingBlock:^(NSNotification *note)
                     {
                         [self description]; // no leak - be careful not to mention self!
                         NSLog(@"%@", @"unregister");
                         [[NSNotificationCenter defaultCenter] removeObserver:observer];
                     }];
}

- (void) viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    // return // uncomment this line so the notification is never posted, and you'll see we leak
    NSLog(@"%@", @"posting");
    [[NSNotificationCenter defaultCenter] postNotificationName:@"woohoo" object:nil];
}

- (void) dealloc {
    NSLog(@"%@", @"dealloc");
}



#endif //====================


#pragma mark - Actions

- (IBAction)done:(id)sender
{
    [self.delegate flipsideViewControllerDidFinish:self];
}

@end

File: bk1ch12p497notificationLeaker/ch12p325NotificationLeaker/FlipsideViewController.swift


import UIKit

protocol FlipsideViewControllerDelegate : class {
    func flipsideViewControllerDidFinish(_ controller:FlipsideViewController)
}

extension Notification.Name {
    static let woohoo = Notification.Name("woohoo")
}

class FlipsideViewController: UIViewController {
    
    weak var delegate : FlipsideViewControllerDelegate!
    
    var observer : Any!
    
    let which = 1
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        switch which {
        case 0:
            self.observer = NotificationCenter.default.addObserver(
            forName: .woohoo, object:nil, queue:nil) {
                _ in
                print("The observer still exists!")
                _ = self.description // leak me, leak me
            }
            // do we still leak if we _don't_ retain the observer? yes
            // (and so does the observer of course)
            self.observer = nil
        case 1:
            self.observer = NotificationCenter.default.addObserver(
            forName: .woohoo, object:nil, queue:nil) {
                [unowned self] _ in // ha ha, fixed it
                print("The observer still exists!")
                _ = self.description
            }
            // do we still leak if we _don't_ retain the observer? no,
            // but the observer has leaked!
            // and now we are in a very dangerous situation...
            // because the observer has a danging pointer to self;
            // we will crash if the notification is posted
            // self.observer = nil
        case 2:
            self.observer = NotificationCenter.default.addObserver(
            forName: .woohoo, object:nil, queue:nil) {
                [weak self] _ in
                print("The observer still exists!")
                // try weak instead
                _ = self?.description
            }
            // now do we still leak if don't retain the observer?
            // no, but the observer has leaked
            // however, at least this is safe because the observer will not do anything
            self.observer = nil
        default:break
        }
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        if let ob = self.observer {
            print("unregister")
            NotificationCenter.default.removeObserver(ob)
        }
    }
    
    @IBAction func done (_ sender: Any!) {
        print("done")
        self.delegate?.flipsideViewControllerDidFinish(self)
    }
    
    // if deinit is not called when you tap Done, we are leaking
    deinit {
        print("deinit")
    }
    
}

extension FlipsideViewController : UIBarPositioningDelegate {
    func position(for bar: UIBarPositioning) -> UIBarPosition {
        return .topAttached
    }
}
